transports/tcp/src/lib.rs

```diff
@@ -22,362 +22,639 @@
+        let domain = if socket_addr.is_ipv4() {
+            Domain::ipv4()
+        } else {
+            Domain::ipv6()
+        };
+        let socket = Socket::new(domain, Type::stream(), Some(socket2::Protocol::tcp()))?;
+        if socket_addr.is_ipv6() {
+            socket.set_only_v6(true)?;
         }
-
-        Ok(Box::pin(do_listen(self, socket_addr).try_flatten_stream()))
+        if let Some(ttl) = self.ttl {
+            socket.set_ttl(ttl)?;
+        }
+        if let Some(nodelay) = self.nodelay {
+            socket.set_nodelay(nodelay)?;
+        }
+        socket.set_reuse_address(true)?;
+        #[cfg(unix)]
+        if let PortReuse::Enabled { .. } = &self.port_reuse {
+            socket.set_reuse_port(true)?;
+        }
+        Ok(socket)
     }
 
-    fn dial(self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
-        let socket_addr =
-            if let Ok(socket_addr) = multiaddr_to_socketaddr(&addr) {
-                if socket_addr.port() == 0 || socket_addr.ip().is_unspecified() {
-                    debug!("Instantly refusing dialing {}, as it is invalid", addr);
-                    return Err(TransportError::Other(io::ErrorKind::ConnectionRefused.into()))
-                }
-                socket_addr
-            } else {
-                return Err(TransportError::MultiaddrNotSupported(addr))
-            };
+    fn do_listen(self, socket_addr: SocketAddr) -> io::Result<TcpListenStream<T>> {
+        let socket = self.create_socket(&socket_addr)?;
+        socket.bind(&socket_addr.into())?;
+        socket.listen(self.backlog as _)?;
+        TcpListenStream::<T>::new(socket.into_tcp_listener(), self.port_reuse)
+    }
 
-        debug!("Dialing {}", addr);
+    async fn do_dial(self, socket_addr: SocketAddr) -> Result<T::Stream, io::Error> {
+        let socket = self.create_socket(&socket_addr)?;
 
-        async fn do_dial(cfg: $tcp_config, socket_addr: SocketAddr) -> Result<$tcp_trans_stream, io::Error> {
-            let stream = <$tcp_stream>::connect(&socket_addr).await?;
-            $apply_config(&cfg, &stream)?;
-            Ok($tcp_trans_stream { inner: stream })
+        if let Some(addr) = self.port_reuse.local_dial_addr(&socket_addr.ip()) {
+            log::trace!("Binding dial socket to listen socket {}", addr);
+            socket.bind(&addr.into())?;
         }
 
-        Ok(Box::pin(do_dial(self, socket_addr)))
+        socket.set_nonblocking(true)?;
+
+        match socket.connect(&socket_addr.into()) {
+            Ok(()) => {}
+            Err(err) if err.raw_os_error() == Some(libc::EINPROGRESS) => {}
+            Err(err) if err.kind() == io::ErrorKind::WouldBlock => {}
+            Err(err) => return Err(err),
+        };
+
+        // let stream = Async::new(socket.into_tcp_stream())?;
+        let stream = T::new_stream(socket.into_tcp_stream()).await?;
+        // stream.writable().await?;
```

Comment:
```suggestion
        let stream = T::new_stream(socket.into_tcp_stream()).await?;
```

For the record `writable().await` is already happening in the corresponding `new_stream` provider implementations.